independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanConfiguratorBase.java

```diff
@@ -230,21 +237,44 @@ public THIS providerType(Type providerType) {
         return self();
     }
 
+    /**
+     * Adds a synthetic injeciton point. The injection point is validated at build time and is also considered when removing
+     * unused beans.
+     *
+     * @param requiredType
+     * @param requiredQualifiers
+     * @return self
+     * @see SyntheticCreationalContext
+     */
+    public THIS addInjectionPoint(Type requiredType, AnnotationInstance... requiredQualifiers) {
+        if (requiredQualifiers.length == 0) {
+            requiredQualifiers = new AnnotationInstance[] { AnnotationInstance.builder(Default.class).build() };
+        }
+        this.injectionPoints.add(new TypeAndQualifiers(requiredType, Set.of(requiredQualifiers)));
+        return self();
+    }
+
     public <U extends T> THIS creator(Class<? extends BeanCreator<U>> creatorClazz) {
         return creator(mc -> {
-            // return new FooBeanCreator().create(context, params)
-            ResultHandle paramsHandle = mc.readInstanceField(
-                    FieldDescriptor.of(mc.getMethodDescriptor().getDeclaringClass(), "params", Map.class),
-                    mc.getThis());
+            // return new FooBeanCreator().create(syntheticCreationalContext)
             ResultHandle creatorHandle = mc.newInstance(MethodDescriptor.ofConstructor(creatorClazz));
-            ResultHandle[] params = { mc.getMethodParam(0), paramsHandle };
             ResultHandle ret = mc.invokeInterfaceMethod(
-                    MethodDescriptor.ofMethod(BeanCreator.class, "create", Object.class, CreationalContext.class, Map.class),
-                    creatorHandle, params);
+                    MethodDescriptor.ofMethod(BeanCreator.class, "create", Object.class, SyntheticCreationalContext.class),
+                    creatorHandle, mc.getMethodParam(1));
             mc.returnValue(ret);
         });
     }
 
+    /**
+     * The first method parameter is the creational context, i.e. the {@code MethodCreator#getMethodParam(0)} returns a
+     * {@link CreationalContext} instance. The second method parameter represents the synthetic creational context, i.e. the
+     * {@code MethodCreator#getMethodParam(1)} returns a {@link SyntheticCreationalContext} instance. Furthermore, the consumer
+     * can read the instance field of name {@code params} and type {@link Map}. This map holds all parameters set via
+     * one of the {@code BeanConfigurator#param()} methods.
```

Comment:
It seems to me that `SyntheticCreationalContext` extends `CreationalContext` and can be used as one. Does it make sense for the method to have 2 parameters, one `CreationalContext` and the other `SyntheticCreationalContext`? Are the two any different?

Also, `SyntheticCreationalContext` has `getParams()` -- calling that is likely preferrable to reading the `params` field? Though I understand we want to retain the `params` field for backwards compatibility.

Reply:
> It seems to me that SyntheticCreationalContext extends CreationalContext and can be used as one. Does it make sense for the method to have 2 parameters, one CreationalContext and the other SyntheticCreationalContext? Are the two any different?

That's for backward compatibility reasons... there might be bytecode creators that rely on this behavior.

>  Though I understand we want to retain the params field for backwards compatibility.

Exactly.

Reply:
Anyway, I believe that most of the users make use of the `BeanCreator` or `SyntheticBeanBuildItem.ExtendedBeanConfigurator`.

Reply:
You mean someone can rely on the fact that the 1st parameter is a `CreationalContext`? If we made the 1st parameter a `SyntheticCreationalContext`, that wouldn't be a breaking change, would it? :thinking:

Reply:
You're right, I'll replace the 1st param. In the first iteration the `SyntheticCreationalContext` interface was called `SyntheticInjectionPoints` and did not implement `CreationalContext` so that's probably the reason why I fixed this signature in my head ;-).

Reply:
When I was going through this, for quite a while, I wanted to suggest to rename `SyntheticCreationalContext` to `SyntheticInjection` :laughing: It might be conceptually easier to understand :shrug:

Reply:
The thing is that having one parameter is much more practical in terms of API evolution...

Reply:
I'm talking about the `BeanCreator`.

Reply:
True, good point.

Reply:
Done.